src/Bundle.ts

```diff
@@ -104,54 +385,115 @@ export default class Bundle {
 		).then(addons => addons.filter(Boolean).join(sep));
 	}
 
-	getPathRelativeToEntryDirname (resolvedId: string): string {
-		if (isRelative(resolvedId) || isAbsolute(resolvedId)) {
-			const entryDirname = dirname(this.entryModule.id);
-			const relativeToEntry = normalize(relative(entryDirname, resolvedId));
+	private setIdentifierRenderResolutions (options: OutputOptions) {
+		const used = blank();
+
+		let dynamicImportMechanism: DynamicImportMechanism;
+		const es = options.format === 'es'
 
-			return isRelative(relativeToEntry)
-				? relativeToEntry
-				: `./${relativeToEntry}`;
+		if (!es) {
+			if (options.format === 'cjs') {
+				dynamicImportMechanism = {
+					left: 'Promise.resolve(require(',
+					right: '))'
+				};
+			} else if (options.format === 'amd') {
+				dynamicImportMechanism = {
+					left: 'new Promise(function (resolve, reject) { require([',
+					right: '], resolve, reject) })'
+				}
+			}
 		}
 
-		return resolvedId;
-	}
+		if (this.graph.dynamicImport) {
+			this.orderedModules.forEach(module => {
+				module.dynamicImportResolutions.forEach((replacement, index) => {
+					const node = module.dynamicImports[index];
 
-	deconflict () {
-		const used = blank();
+					if (!replacement)
+						return;
+
+					if (replacement instanceof Module) {
+						// if we have the module in the bundle, inline as Promise.resolve(namespace)
+						// ensuring that we create a namespace import of it as well
+						if (replacement.bundle === this) {
+							node.setResolution(replacement.namespace(), { left: 'Promise.resolve().then(() => ', right: ')' });
+						// for the module in another chunk, import that other chunk directly
+						} else {
+							node.setResolution(`"${replacement.bundle.id}"`, dynamicImportMechanism);
+						}
+					// external dynamic import resolution
+					} else if (replacement instanceof ExternalModule) {
+						node.setResolution(`"${replacement.id}"`, dynamicImportMechanism);
+					// AST Node -> source replacement
+					} else {
+						node.setResolution(replacement, dynamicImportMechanism);
+					}
+				});
+			});
+		}
 
 		// ensure no conflicts with globals
 		keys(this.graph.scope.variables).forEach(name => (used[name] = 1));
 
 		function getSafeName (name: string): string {
```

Comment:
If we supply `used`, this function could be extracted out of this class. Also, this function might then be reused to deconflict other things like bundle file names.

Reply:
That sounds like a very sensible refactoring to do, I was aiming to just leave the utility functions exactly as-is for this PR. Further PR work can definitely dive into some of these refactorings.

Reply:
👍